planning/behavior_velocity_planner/src/scene_module/invalid_lanelet/scene.cpp

```diff
@@ -0,0 +1,280 @@
+// Copyright 2023 TIER IV, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "scene_module/invalid_lanelet/scene.hpp"
+
+#include "utilization/util.hpp"
+
+#include <tier4_autoware_utils/tier4_autoware_utils.hpp>
+
+#include <lanelet2_core/utility/Optional.h>
+
+namespace behavior_velocity_planner
+{
+using tier4_autoware_utils::createPoint;
+namespace bg = boost::geometry;
+
+InvalidLaneletModule::InvalidLaneletModule(
+  const int64_t module_id, const int64_t lane_id, const PlannerParam & planner_param,
+  const rclcpp::Logger logger, const rclcpp::Clock::SharedPtr clock)
+: SceneModuleInterface(module_id, logger, clock),
+  lane_id_(lane_id),
+  planner_param_(planner_param),
+  state_(State::INIT)
+{
+  velocity_factor_.init(VelocityFactor::INVALID_LANELET);
+}
+
+bool InvalidLaneletModule::modifyPathVelocity(PathWithLaneId * path, StopReason * stop_reason)
```

Comment:
Would you think of simplification for this function? I guess it can be split into several functions.

Reply:
Thanks @TakaHoribe san for your review and providing the valuable comments.

For the `modifyPathVelocity` function simplification, I have worked on implementation refactoring and I hope it looks now better. Please have a look and let me know your thoughts.


For the the behavior of `stop line moves following the vehicle`, let me explain my understanding ...

- my understanding to the behavior of invalid lanelet module when it comes to `stopped state`, it should not go out of this state until the human driver takes over the driving responsibility and the autoware will be re-engaged (hence, re-initialized) and the virtual wall (stop line) will not be there as long as we are outside of an invalid lanelet after re-engaging . 

- As long as we are in an invalid lanelet, it is true that the stop pose is exactly in front of the vehicle so that it will prevent it from moving if the human driver re-engaged the autonomous mode.

- For the case of suddenly locating the vehicle in a new different start pose without turning off the system and turning on again, I guess this is not a real life situation. Do I need to handle this scenario or consider that autoware will be re-initialized in such case ?

Please correct my understanding about the above points if it is not true, and according to our discussion we can go forward if any changes needed.

Reply:
> As long as we are in an invalid lanelet, it is true that the stop pose is exactly in front of the vehicle so that it will prevent it from moving if the human driver re-engaged the autonomous mode.

@ahmeddesokyebrahim Thank you, I have the same opinion.

What I tried is "after the vehicle is stopped by `invalid lanelet`, change to manual operation, move the vehicle backward, and then re-engage".
In this case, after the vehicle was moved backward up by manual driving and exited the `invalid lanelet`, it still continued to stop and could not be started by re-engaging. 
We can keep the stopline in front of the ego vehicle when it is in the `invalid lanelet`, but the stopline should not be out of the `invalid lanelet`.

Well, this is not a big issue and can be solved/discussed out of this PR. Let's merge the PR.